src/metabase/search/filter.clj

```diff
@@ -0,0 +1,112 @@
+   [metabase.driver.common.parameters.dates :as params.dates]
+   [metabase.public-settings.premium-features :as premium-features]
+   [metabase.search.in-place.filter :as search.in-place.filter]
+   [metabase.search.spec :as search.spec]
+   [metabase.util :as u]
+   [metabase.util.date-2 :as u.date]
+   [metabase.util.i18n :refer [tru]])
+  (:import
+   (java.time LocalDate)))
+
+(def ^:private context-key->attr
+  {:archived?           :archived
+   :created-at          :created-at
+   :created-by          :creator-id
+   :last-edited-at      :last-edited-at
+   :last-edited-by      :last-editor-id
+   :search-native-query :dataset-query
+   ;; this actually has nothing to do with tables anymore, as we also filter cards.
+   :table-db-id         :database-id
+   :verified            :verified})
+
+(def ^:private attr->index-key
+  (into {} (for [k (vals context-key->attr)]
+             [k (keyword (str "search_index." (u/->snake_case_en (name k))))])))
+
+(defn search-context->applicable-models
+  "Returns a set of models that are applicable given the search context.
+
+  If the context has optional filters, the models will be restricted for the set of supported models only."
+  [search-context]
+  (if (= :search-engine/in-place (:search-engine search-context))
+    (search.in-place.filter/search-context->applicable-models search-context)
+    (let [required (keep context-key->attr (keys search-context))]
+      (into #{}
+            (remove nil?)
+            (for [search-model (:models search-context)
+                  :let [spec (search.spec/spec search-model)]]
+              (when (every? (:attrs spec) required)
+                (:name spec)))))))
+
+(defn- date-range-filter-clause
+  [dt-col dt-val]
+  (let [date-range (try
+                     (params.dates/date-string->range dt-val {:inclusive-end? false})
+                     (catch Exception _e
+                       (throw (ex-info (tru "Failed to parse datetime value: {0}" dt-val) {:status-code 400}))))
+        start      (some-> (:start date-range) u.date/parse)
+        end        (some-> (:end date-range) u.date/parse)
+        dt-col     (if (some #(instance? LocalDate %) [start end])
+                     [:cast dt-col :date]
+                     dt-col)]
+    (cond
+      (= start end)
+      [:= dt-col start]
+
+      (nil? start)
+      [:< dt-col end]
+
+      (nil? end)
+      [:> dt-col start]
+
+      :else
+      [:and [:>= dt-col start] [:< dt-col end]])))
+
+(defmulti ^:private where-clause* (fn [context-key _column _v] context-key))
+
+(defmethod where-clause* :archived? [_ k v] [:= k v])
+
+(defmethod where-clause* :created-at [_ k v] (date-range-filter-clause k v))
+
+(defmethod where-clause* :created-by [_ k v] [:in k v])
+
+(defmethod where-clause* :last-edited-at [_ k v] (date-range-filter-clause k v))
+
+(defmethod where-clause* :last-edited-by [_ k v] [:in k v])
+
+(defmethod where-clause* :table-db-id [_ k v] [:= k v])
```

Comment:
So this looks like we're going with separate fields for filters, right?

Reply:
Yeah for now I'm thinking fields which we can index on, this is the lowest common denominator for deprecating legacy.

We can experiment with a vector for the categorical ones as a final step. I know I know, facets 😄